include/graphblas/algorithms/conjugate_gradient.hpp

```diff
@@ -323,110 +375,211 @@ namespace grb {
 			} else {
 				ret = ret ? ret : grb::dot< descr_dense >( bnorm, b, b, ring );
 			}
-			assert( ret == SUCCESS );
+			assert( ret == grb::SUCCESS );
 
-			if( ret == SUCCESS ) {
+			// get effective tolerance and exit on any error during prelude
+			if( ret == grb::SUCCESS ) {
 				tol *= std::sqrt( grb::utils::is_complex< IOType >::modulus( bnorm ) );
+			} else {
+				return ret;
 			}
 
+			// all OK; perform main iterations
 			size_t iter = 0;
-
 			do {
 				assert( iter < max_iterations );
 				(void) ++iter;
 
 				// temp = 0
 				ret = ret ? ret : grb::set( temp, 0 );
-				assert( ret == SUCCESS );
+				assert( ret == grb::SUCCESS );
 
 				// temp = A * u;
 				ret = ret ? ret : grb::mxv< descr_dense >( temp, A, u, ring );
-				assert( ret == SUCCESS );
+				assert( ret == grb::SUCCESS );
 
-				// beta = u' * temp
+				// beta = temp' * u
 				beta = zero;
 				if( grb::utils::is_complex< IOType >::value ) {
```

Comment:
You could avoid `is_complex` check by passing  _pass _conjugate_right_mul_ operator_ to `dot()` call, i.e. take a look in at [this](https://github.com/Algebraic-Programming/ALP/blob/68b6256e2ac0672d648980487399c9d18693fe59/include/graphblas/algorithms/gmres.hpp#L556C1-L561C9) section.

Reply:
Very elegant indeed -- I applied it